Core/Parsing/Structure/IntermediateModel/SupportedMethodSpecifications.cs

```diff
@@ -96,12 +96,27 @@ private static bool HasIndexSelectorParameter (MethodInfo methodInfo, int parame
       if (parameters.Length <= parameterPosition)
         return false;
 
-      var selectorType = parameters[parameterPosition].ParameterType.GetTypeInfo();
-      // Enumerable taks a Func<...> but Querable takes an Expression<Func<...>>
-      if (typeof (Expression).GetTypeInfo().IsAssignableFrom (selectorType))
-        selectorType = selectorType.GenericTypeArguments[0].GetTypeInfo();
+      var selectorType = parameters[parameterPosition].ParameterType.GetTypeInfo().UnwrapEnumerable();
 
       return selectorType.GenericTypeArguments[1] == typeof (int);
     }
+
+    private static TypeInfo UnwrapEnumerable (this TypeInfo typeInfo)
+    {
+      var comparedTypeInfo = typeInfo;
+
+#if !NET_4_0 && !NET_3_5
+      // Workaround for a corner case in .Net Native because
+      // IsAssignableFrom throws for open generics.
+      if (typeInfo.ContainsGenericParameters)
+        comparedTypeInfo = typeInfo.BaseType.GetTypeInfo();
```

Comment:
Would we ever need to do this more than once until we find a base type that isn't generic?